### PR TITLE
Countdown and staircase timers. DHCP hostname

### PR DESCRIPTION
**Pull request by klew**

### CODE REVIEW

File: src/user/supla_esp_cfgmode.c [hunk 1]
```diff
@@ -877,29 +906,11 @@ void ICACHE_FLASH_ATTR supla_esp_cfgmode_start(void) {
...
-
-  if (apssid_len + mac_str_len > sizeof(apconfig.ssid)) {
-    apssid_len -= apssid_len + mac_str_len - sizeof(apconfig.ssid);
-  }
+  char ssid_name[sizeof(apconfig.ssid)] = {};
+  int ssid_name_length = supla_esp_cfgmode_generate_ssid_name(ssid_name, sizeof(apconfig.ssid));
 
-  memcpy(&apconfig.ssid[apssid_len], mac_str, mac_str_len);
-  apconfig.ssid_len = apssid_len + mac_str_len;
+  memcpy(&apconfig.ssid, ssid_name, ssid_name_length);
```
Comment (przemyslawzygmunt) [inline src/user/supla_esp_cfgmode.c:912]: &apconfig.ssid pointer to pointer ?

File: src/user/supla_esp_cfgmode.c [hunk 2]
```diff
@@ -877,29 +906,11 @@ void ICACHE_FLASH_ATTR supla_esp_cfgmode_start(void) {
...
-               (unsigned char)mac[4], (unsigned char)mac[5]);
-#endif /*CFGMODE_SSID_LIMIT_MACLEN*/
-
-  int mac_str_len = strnlen(mac_str, sizeof(mac_str));
-
-  if (apssid_len + mac_str_len > sizeof(apconfig.ssid)) {
-    apssid_len -= apssid_len + mac_str_len - sizeof(apconfig.ssid);
-  }
+  char ssid_name[sizeof(apconfig.ssid)] = {};
+  int ssid_name_length = supla_esp_cfgmode_generate_ssid_name(ssid_name, sizeof(apconfig.ssid));
```
Comment (przemyslawzygmunt) [inline src/user/supla_esp_cfgmode.c:910]: supla_esp_cfgmode_generate_ssid_name(apconfig.ssid, sizeof(apconfig.ssid)); ?

File: src/user/supla_esp_gpio.h [hunk 3]
```diff
@@ -136,13 +135,18 @@ void GPIO_ICACHE_FLASH supla_esp_gpio_state_update(void);
 
 void supla_esp_gpio_start_input_timer(supla_input_cfg_t *input_cfg);
 
-void supla_esp_gpio_set_hi(int port, char hi);
+void supla_esp_gpio_set_hi(int port, unsigned char hi);
```
Comment (przemyslawzygmunt) [inline src/user/supla_esp_gpio.h:138]: I'm not sure if switching to unsigned will break something somewhere.

**Review by przemyslawzygmunt — CHANGES_REQUESTED** [review]
I think apconfig.ssid should be instead of &apconfig.ssid

File: src/user/supla_esp_gpio.h (hunk 3 above)
Comment (klew) [reply src/user/supla_esp_gpio.h:138]: we are using 255 in code for "toggle". "char" may be signed or unsigned - this is not defined.
With "char" UT tests were not working properly with 255 value passed.

File: src/user/supla_esp_cfgmode.c (hunk 1 above)
Comment (klew) [reply src/user/supla_esp_cfgmode.c]: fixed

File: src/user/supla_esp_cfgmode.c (hunk 2 above)
Comment (klew) [reply src/user/supla_esp_cfgmode.c]: right, thanks!

**Review by przemyslawzygmunt — APPROVED** [review]

**Pull request merged**